File: apps/openwall/src/camera.ts
/**
 * Turning a tap on a photograph into a ray in world space, and back again.
 *
 * ARKit gives each frame a camera transform and intrinsics. Those two things are
 * what make a photograph of a stud bay into a measurement rather than a picture.
 */

import {
  type Mat4, type Vec3, invertRigid, normalise, transformDirection,
  translationOf, transformPoint, vec,
} from "./vec.js";

/** Pinhole intrinsics in pixels, as ARKit reports them. */
export interface Intrinsics {
  readonly fx: number;
  readonly fy: number;
  readonly cx: number;
  readonly cy: number;
}

export interface CameraFrame {
  readonly id: string;
  /** Camera-to-world, column-major, ARKit convention (-Z forward, +Y up). */
  readonly transform: Mat4;
  readonly intrinsics: Intrinsics;
  readonly imageWidth: number;
  readonly imageHeight: number;
}

export interface Ray {
  readonly origin: Vec3;
  /** Unit length. */
  readonly direction: Vec3;
}

/** A tap on a photograph, in pixels, origin top-left. */
export interface Tap {
  readonly frame: CameraFrame;
  readonly x: number;
  readonly y: number;
}

function assertIntrinsics({ fx, fy }: Intrinsics): void {
  if (fx === 0 || fy === 0) {
    throw new RangeError("focal length cannot be zero");
  }
}

/**
 * Unproject a pixel into a world-space ray.
 *
 * Image y runs down and camera y runs up, so the y term is negated. Camera z is
 * -1 because ARKit's camera looks down its own -Z.
 */
export function rayThroughPixel(frame: CameraFrame, px: number, py: number): Ray {
  assertIntrinsics(frame.intrinsics);
  const { fx, fy, cx, cy } = frame.intrinsics;
  const inCamera = vec((px - cx) / fx, -(py - cy) / fy, -1);
  return {
    origin: translationOf(frame.transform),
    direction: normalise(transformDirection(frame.transform, inCamera)),
  };
}

export const rayThroughTap = (tap: Tap): Ray => rayThroughPixel(tap.frame, tap.x, tap.y);

export interface Projection {
  readonly x: number;
  readonly y: number;
  /** False when the point is behind the camera, where the maths still yields numbers. */
  readonly inFront: boolean;
  /** True when the pixel also falls inside the image bounds. */
  readonly onImage: boolean;
}

/**
 * Project a world point back onto a frame - what draws a traced run over the
 * photograph it was traced in.
 */
export function projectToPixel(frame: CameraFrame, world: Vec3): Projection {
  assertIntrinsics(frame.intrinsics);
  const { fx, fy, cx, cy } = frame.intrinsics;
  const p = transformPoint(invertRigid(frame.transform), world);
  const depth = -p.z;
  if (depth === 0) {
    return { x: Number.NaN, y: Number.NaN, inFront: false, onImage: false };
  }
  const x = (fx * p.x) / depth + cx;
  const y = cy - (fy * p.y) / depth;
  const inFront = depth > 0;
  return {
    x, y, inFront,
    onImage:
      inFront && x >= 0 && y >= 0 && x <= frame.imageWidth && y <= frame.imageHeight,
  };
}

File: apps/MeasureKit/Sources/MeasureKit/Camera.swift
import Foundation

/// Pinhole intrinsics in pixels, as ARKit reports them.
public struct Intrinsics: Sendable {
    public var fx: Double, fy: Double, cx: Double, cy: Double
    public init(fx: Double, fy: Double, cx: Double, cy: Double) {
        precondition(fx != 0 && fy != 0, "focal length cannot be zero")
        self.fx = fx; self.fy = fy; self.cx = cx; self.cy = cy
    }
}

public struct CameraFrame: Sendable {
    public var id: String
    /// Camera-to-world, ARKit convention (-Z forward, +Y up).
    public var transform: Transform
    public var intrinsics: Intrinsics
    public var imageWidth: Double
    public var imageHeight: Double

    public init(id: String, transform: Transform, intrinsics: Intrinsics,
                imageWidth: Double, imageHeight: Double) {
        self.id = id; self.transform = transform; self.intrinsics = intrinsics
        self.imageWidth = imageWidth; self.imageHeight = imageHeight
    }
}

public struct Ray: Sendable {
    public var origin: Vector3
    /// Unit length.
    public var direction: Vector3
    public init(origin: Vector3, direction: Vector3) {
        self.origin = origin; self.direction = direction
    }
}

/// A tap on a photograph, in pixels, origin top-left.
public struct Tap: Sendable {
    public var frame: CameraFrame
    public var x: Double, y: Double
    public init(frame: CameraFrame, x: Double, y: Double) {
        self.frame = frame; self.x = x; self.y = y
    }
    public var ray: Ray { frame.ray(throughX: x, y: y) }
}

public extension CameraFrame {
    /// Unproject a pixel into a world-space ray.
    ///
    /// Image y runs down and camera y runs up, so the y term is negated. Camera z
    /// is -1 because ARKit's camera looks along its own -Z.
    func ray(throughX px: Double, y py: Double) -> Ray {
        let inCamera = Vector3((px - intrinsics.cx) / intrinsics.fx,
                               -(py - intrinsics.cy) / intrinsics.fy,
                               -1)
        return Ray(origin: transform.translation,
                   direction: transform.direction(inCamera).normalised())
    }

    /// Project a world point back onto this frame - what draws a traced run over
    /// the photograph it was traced in.
    func project(_ world: Vector3) -> Projection {
        let p = transform.invertedRigid().point(world)
        let depth = -p.z
        guard depth != 0 else {
            return Projection(x: .nan, y: .nan, isInFront: false, isOnImage: false)
        }
        let x = intrinsics.fx * p.x / depth + intrinsics.cx
        let y = intrinsics.cy - intrinsics.fy * p.y / depth
        let inFront = depth > 0
        return Projection(x: x, y: y, isInFront: inFront,
                          isOnImage: inFront && x >= 0 && y >= 0
                                     && x <= imageWidth && y <= imageHeight)
    }
}

public struct Projection: Sendable {
    public var x: Double, y: Double
    /// False when the point is behind the camera, where the maths still yields numbers.
    public var isInFront: Bool
    public var isOnImage: Bool
}
